### ISSUE TITLE
REQUEST: Create `kubernetes/cri` staging repository

### ISSUE THREAD

**Issue by saschagrunert** [kubernetes/org#4805]
### New repo, staging repo, or migrate existing

new repo

### Is it a staging repo?

yes

### Requested name for new repository

cri

### Which Organization should it reside

kubernetes

### Who should have admin access?

SIG Node admins

### Who should have write access?

SIG Node admins and the publishing bot

### Who should be listed as approvers in OWNERS?

SIG Node Approvers

### Who should be listed in SECURITY_CONTACTS?

committee-security-response

### What should the repo description be?

Container Runtime Interface client implementation

### What SIG and subproject does this fall under?

SIG Node

### Please provide references to appropriate approval for this new repo

- Discussion: https://groups.google.com/g/kubernetes-sig-architecture/c/RZH8bY5ONy4/m/MHny9WuYAQAJ

cc @SIGNODE

**Comment by dims** [comment]
+1 from me

**Comment by SergeyKanzhelev** [comment]
I like the idea of exposing the opinionated client library for CRI. Definitely would have helped when we migrated from ``v1alpha2` to `v1` and had the fallback logic implemented in both places - crictl and kubelet. Expanding a little bit on this for everybody's context.

So the client will have the following features:
- choosing the right API version and establishing connection
- check the service connectivity (`validateServiceConnection`)
- gRPC options including various timeouts and the fact that some calls do have timeouts and others - don't
- Tracing 
- Error handling

It may be a good idea to move the instrumented_service as well to also unify on metric names from those calls.

The client will not change any promises regarding which runtime versions skews we support. Also no promises on back compatibility of golang APIs as it will be versioned as `0.x`. Making it public puts a little bit of pressure on maintainers to minimize the breaking changes in golang APIs. But I don't remember many changes there recently, so it is not a big issue.

A few comments on some specifics:

1. Is cri-client will be a better name?
2. Do we need to remove all the logs from the library? It may not be expected by clients to see those logs.
3. As mentioned above, it may be a good idea to move the instrumented_service as well.
4. We have a comment on error handling that we may not need specific TImeoutError type. I think it is coming from the time when we had both - dockershim and CRI. If we decide to clean it up - it will be a golang API change. We need to be open to those kind of changes.

**Comment by saschagrunert** [comment]
> A few comments on some specifics:
> 
> 1. Is cri-client will be a better name?

Yes probably, what do others think?

> 2. Do we need to remove all the logs from the library? It may not be expected by clients to see those logs.

Yeah we could make the logging an optional part before moving to code over to staging.

> 3. As mentioned above, it may be a good idea to move the instrumented_service as well.

Yes, what about utilities like log parsing?

https://github.com/kubernetes/kubernetes/blob/7ea3d0245a63fbbba698f1cb939831fe8143db3e/pkg/kubelet/kuberuntime/logs/logs.go#L283

> 4. We have a comment on error handling that we may not need specific TImeoutError type. I think it is coming from the time when we had both - dockershim and CRI. If we decide to clean it up - it will be a golang API change. We need to be open to those kind of changes.

Means we should remove that type before going to staging?

**Comment by saschagrunert** [comment]
We can also move the cri streaming server implementation there: https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/kubelet/pkg/cri/streaming

So I'd say we can just stick to `cri`.

**Comment by mrbobbytables** [comment]
just wanted to check in on the status of this. Is this still a thing?

**Comment by saschagrunert** [comment]
> just wanted to check in on the status of this. Is this still a thing?

Yes, we just wait on consensus from SIG Node. @mrunalp @haircommander do we have an update from that side?

**Comment by SergeyKanzhelev** [comment]
I am generally supportive. My suggestion is to define what will be in scope of the library and when we do clean ups like removing logs - before or after moving it to staging.

As for including the streaming server to the same library. @saschagrunert do you envision crictl reusing some of the methods? It is still a client, right? I am just not sure the `cri` as a name of a package is very descriptive and if many people may get confused

**Comment by saschagrunert** [comment]
Changed the name to `cri-client` for now.

**Comment by SergeyKanzhelev** [comment]
I am happy with everything here! I think it will be a great library to expose and will improve all the cri clients

**Comment by mrunalp** [comment]
:+1: from me.

**Comment by cblecker** [comment]
/assign

I've created the repo and the blank commit, as well as approved https://github.com/kubernetes/org/pull/4806

Next steps:
- The staging repo content needs to be merged into https://github.com/kubernetes/kubernetes/tree/master/staging -- note this should include standard templated content (https://github.com/kubernetes/kubernetes-template-project) and things like the pull request template to deter direct PRs
- https://github.com/kubernetes/test-infra/pull/32537 needs to merge
- Once all the above is done, it needs to be added to https://github.com/kubernetes/publishing-bot/blob/master/hack/repos.sh
- Added to sigs.yaml

@saschagrunert: will you be taking care of these?

**Comment by saschagrunert** [comment]
> @saschagrunert: will you be taking care of these?

Yes, following-up in https://github.com/kubernetes/kubernetes/pull/123797

**Comment by saschagrunert** [comment]
Looks like we’re don, thank y’all for the support!
